File: app/src/main/java/org/neidhardt/dynamicsoundboard/fileexplorer/LoadLayoutDialog.java
package org.neidhardt.dynamicsoundboard.fileexplorer;

import android.annotation.SuppressLint;
import android.app.Dialog;
import android.app.FragmentManager;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v7.app.AppCompatDialog;
import android.support.v7.widget.DefaultItemAnimator;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.widget.Toast;
import org.neidhardt.dynamicsoundboard.R;
import org.neidhardt.dynamicsoundboard.dao.MediaPlayerData;
import org.neidhardt.dynamicsoundboard.dao.SoundSheet;
import org.neidhardt.dynamicsoundboard.mediaplayer.EnhancedMediaPlayer;
import org.neidhardt.dynamicsoundboard.misc.JsonPojo;
import org.neidhardt.dynamicsoundboard.views.recyclerviewhelpers.DividerItemDecoration;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * File created by eric.neidhardt on 14.11.2014.
 */
public class LoadLayoutDialog extends FileExplorerDialog implements View.OnClickListener
{
	private static final String TAG = LoadLayoutDialog.class.getName();

	private View confirm;
	private RecyclerView directories;

	public static void showInstance(FragmentManager manager)
	{
		LoadLayoutDialog dialog = new LoadLayoutDialog();
		dialog.show(manager, TAG);
	}

	@NonNull
	@Override
	public Dialog onCreateDialog(Bundle savedInstanceState)
	{
		@SuppressLint("InflateParams") View view = this.getActivity().getLayoutInflater().inflate(R.layout.dialog_load_sound_sheets, null);
		view.findViewById(R.id.b_cancel).setOnClickListener(this);
		this.confirm = view.findViewById(R.id.b_ok);
		this.confirm.setOnClickListener(this);
		this.confirm.setEnabled(false);

		this.directories = (RecyclerView)view.findViewById(R.id.rv_dialog);
		this.directories.addItemDecoration(new DividerItemDecoration());
		this.directories.setLayoutManager(new LinearLayoutManager(this.getActivity()));
		this.directories.setItemAnimator(new DefaultItemAnimator());
		this.directories.setAdapter(super.getAdapter());

		String previousPath = this.getPathFromSharedPreferences(TAG);
		if (previousPath != null)
			super.getAdapter().setParent(new File(previousPath));

		AppCompatDialog dialog = new AppCompatDialog(this.getActivity(), R.style.DialogThemeNoTitle);
		dialog.setContentView(view);

		return dialog;
	}

	@Override
	protected void onFileSelected()
	{
		this.confirm.setEnabled(true);
		int position = super.getAdapter().getFileList().indexOf(super.getAdapter().getSelectedFile());
		this.directories.scrollToPosition(position);
	}

	@Override
	protected boolean canSelectDirectory()
	{
		return false;
	}

	@Override
	protected boolean canSelectFile()
	{
		return true;
	}

	@Override
	public void onClick(@NonNull View v)
	{
		switch (v.getId())
		{
			case R.id.b_cancel:
				this.dismiss();
				break;
			case R.id.b_ok:
				File currentDirectory = super.getAdapter().getParentFile();
				if (currentDirectory != null)
					this.storePathToSharedPreferences(TAG, currentDirectory.getPath());

				if (super.getAdapter().getSelectedFile() != null)
					this.loadFromFileAndDismiss(super.getAdapter().getSelectedFile());
				else
					Toast.makeText(this.getActivity(), R.string.dialog_load_layout_no_file_info, Toast.LENGTH_SHORT).show();
				break;
		}
	}

	private void loadFromFileAndDismiss(File file)
	{
		try
		{
			JsonPojo parsedJson = JsonPojo.readFromFile(file);

			List<SoundSheet> soundSheets = parsedJson.getSoundSheets();
			List<MediaPlayerData> playList = parsedJson.getPlayList();
			Map<String, List<MediaPlayerData>> sounds = parsedJson.getSounds();

			this.addLoadedSoundSheets(soundSheets);

			this.addLoadedSounds(sounds);
			this.addLoadedPlayList(playList);

			this.dismiss();
		}
		catch (IOException e)
		{
			e.printStackTrace();
		}
	}

	private void addLoadedSoundSheets(List<SoundSheet> newSoundSheets)
	{
		List<SoundSheet> oldCurrentSoundSheet = this.getSoundSheetsDataAccess().getSoundSheets();

		List<EnhancedMediaPlayer> playersToRemove = new ArrayList<>();
		for (SoundSheet soundSheet : oldCurrentSoundSheet)
			playersToRemove.addAll(this.getSoundsDataAccess().getSoundsInFragment(soundSheet.getFragmentTag()));

		this.getSoundsDataStorage().removeSounds(playersToRemove);
		this.getSoundSheetsDataStorage().removeSoundSheets(oldCurrentSoundSheet);


		for (SoundSheet soundSheet : newSoundSheets)
			this.getSoundSheetsDataStorage().addSoundSheetToManager(soundSheet);
	}

	private void addLoadedPlayList(List<MediaPlayerData> playList)
	{
		this.getSoundsDataStorage().removeSoundsFromPlaylist(this.getSoundsDataAccess().getPlaylist()); // clear playlist before adding new values

		for (MediaPlayerData mediaPlayerData : playList)
			this.getSoundsDataStorage().createPlaylistSoundAndAddToManager(mediaPlayerData);
	}

	private void addLoadedSounds(Map<String, List<MediaPlayerData>> sounds)
	{
		for (String key : sounds.keySet())
		{
			List<MediaPlayerData> soundsPerFragment = sounds.get(key);
			for (MediaPlayerData mediaPlayerData : soundsPerFragment)
				this.getSoundsDataStorage().createSoundAndAddToManager(mediaPlayerData);
		}
	}

}

File: app/src/main/kotlin/org/neidhardt/dynamicsoundboard/fileexplorer/LoadLayoutDialog.kt
package org.neidhardt.dynamicsoundboard.fileexplorer

import android.annotation.SuppressLint
import android.app.Dialog
import android.os.Bundle
import android.support.v4.app.FragmentManager
import android.support.v7.app.AlertDialog
import android.support.v7.widget.DefaultItemAnimator
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.widget.Toast
import org.neidhardt.dynamicsoundboard.R
import org.neidhardt.dynamicsoundboard.SoundboardApplication
import org.neidhardt.dynamicsoundboard.dao.MediaPlayerData
import org.neidhardt.dynamicsoundboard.dao.SoundSheet
import org.neidhardt.dynamicsoundboard.mediaplayer.MediaPlayerController
import org.neidhardt.dynamicsoundboard.misc.readFromFile
import org.neidhardt.dynamicsoundboard.reportError
import org.neidhardt.android_utils.recyclerview_utils.decoration.DividerItemDecoration
import java.io.File
import java.io.IOException
import java.util.*

/**
 * File created by eric.neidhardt on 14.11.2014.
 */
class LoadLayoutDialog : FileExplorerDialog(), LayoutStorageDialog
{
	private var directories: RecyclerView? = null

	private val soundSheetsDataAccess = SoundboardApplication.soundSheetsDataAccess
	private val soundSheetsDataStorage = SoundboardApplication.soundSheetsDataStorage
	private val soundsDataAccess = SoundboardApplication.soundsDataAccess
	private val soundsDataStorage = SoundboardApplication.soundsDataStorage

	companion object {
		private val TAG = LoadLayoutDialog::class.java.name

		fun showInstance(manager: FragmentManager)
		{
			val dialog = LoadLayoutDialog()
			dialog.show(manager, TAG)
		}
	}

	override fun onCreateDialog(savedInstanceState: Bundle?): Dialog
	{
		@SuppressLint("InflateParams") val view = this.activity.layoutInflater.inflate(R.layout.dialog_load_sound_sheets, null)

		this.directories = (view.findViewById(R.id.rv_dialog) as RecyclerView).apply {
			this.addItemDecoration(DividerItemDecoration(this.context, R.color.background, R.color.divider))
			this.layoutManager = LinearLayoutManager(this.context)
			this.itemAnimator = DefaultItemAnimator()
		}
		this.directories?.adapter = super.adapter

		val previousPath = this.getPathFromSharedPreferences(LayoutStorageDialog.KEY_PATH_STORAGE)
		if (previousPath != null)
			super.adapter.setParent(File(previousPath))

		return AlertDialog.Builder(this.activity).apply {
			this.setView(view)
			this.setNegativeButton(R.string.dialog_cancel, { dialogInterface, i -> dismiss() })
			this.setPositiveButton(R.string.dialog_load, { dialogInterface, i -> onConfirm() })
		}.create()
	}

	override fun onFileSelected(selectedFile: File)
	{
		val position = super.adapter.fileList.indexOf(selectedFile)
		this.directories!!.scrollToPosition(position)
	}

	override fun canSelectMultipleFiles(): Boolean = false

	override fun canSelectDirectory(): Boolean = false

	override fun canSelectFile(): Boolean = true

	private fun onConfirm()
	{
		val currentDirectory = super.adapter.parentFile
		if (currentDirectory != null)
			this.storePathToSharedPreferences(LayoutStorageDialog.KEY_PATH_STORAGE, currentDirectory.path)

		if (super.adapter.selectedFiles.size != 0)
			this.loadFromFileAndDismiss(super.adapter.selectedFiles.iterator().next())
		else
			Toast.makeText(this.activity, R.string.dialog_load_layout_no_file_info, Toast.LENGTH_SHORT).show()
	}

	private fun loadFromFileAndDismiss(file: File)
	{
		try
		{
			val parsedJson = readFromFile(file)

			val soundSheets = parsedJson.soundSheets
			val playList = parsedJson.playList
			val sounds = parsedJson.sounds

			this.addLoadedSoundSheets(soundSheets)

			this.addLoadedSounds(sounds)
			this.addLoadedPlayList(playList)

			this.dismiss()
		} catch (e: IOException)
		{
			e.printStackTrace()
			SoundboardApplication.reportError(e)
		}
	}

	private fun addLoadedSoundSheets(newSoundSheets: List<SoundSheet>)
	{
		val oldCurrentSoundSheet = this.soundSheetsDataAccess.getSoundSheets()

		val playersToRemove = ArrayList<MediaPlayerController>()
		for (soundSheet in oldCurrentSoundSheet)
			playersToRemove.addAll(this.soundsDataAccess.getSoundsInFragment(soundSheet.fragmentTag))

		this.soundsDataStorage.removeSounds(playersToRemove)
		this.soundSheetsDataStorage.removeSoundSheets(oldCurrentSoundSheet)

		for (soundSheet in newSoundSheets)
			this.soundSheetsDataStorage.addSoundSheetToManager(soundSheet)
	}

	private fun addLoadedPlayList(playList: List<MediaPlayerData>)
	{
		this.soundsDataStorage.removeSoundsFromPlaylist(this.soundsDataAccess.playlist) // clear playlist before adding new values

		for (mediaPlayerData in playList)
			this.soundsDataStorage.createPlaylistSoundAndAddToManager(mediaPlayerData)
	}

	private fun addLoadedSounds(sounds: Map<String, List<MediaPlayerData>>)
	{
		for (key in sounds.keys) {
			val soundsPerFragment = sounds[key].orEmpty()
			for (mediaPlayerData in soundsPerFragment)
				this.soundsDataStorage.createSoundAndAddToManager(mediaPlayerData)
		}
	}

}
